var/spack/repos/builtin/packages/r-rpostgres/package.py

```diff
@@ -0,0 +1,39 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class RRpostgres(RPackage):
+    """'Rcpp' Interface to 'PostgreSQL'
+
+    Fully 'DBI'-compliant 'Rcpp'-backed interface to 'PostgreSQL'
+    <https://www.postgresql.org/>, an open-source relational database."""
+
+    homepage = "https://rpostgres.r-dbi.org/"
+    cran     = "RPostgres"
+
+    version('1.3.1', sha256='f68ab095567317ec32d3faa10e5bcac400aee5aeca8d7132260d4e90f82158ea')
+
+    depends_on('r@3.1.0:', type=('build', 'run'))
+    depends_on('r-bit64', type=('build', 'run'))
+    depends_on('r-blob@1.2.0:', type=('build', 'run'))
+    depends_on('r-dbi@1.1.0:', type=('build', 'run'))
+    depends_on('r-hms@0.5.0:', type=('build', 'run'))
+    depends_on('r-lubridate', type=('build', 'run'))
+    depends_on('r-rcpp@0.11.4.2:', type=('build', 'run'))
+    depends_on('r-withr', type=('build', 'run'))
+    depends_on('r-bh', type=('build', 'run'))
+    depends_on('r-plogr@0.2.0:', type=('build', 'run'))
+    depends_on('postgresql@9.0:', type=('build', 'run'))
+
+    patch('configure_add_rpath.patch')
```

Comment:
Can you add a very brief comment above the directive (the name of the file is explicative, but having also one line saying that this is needed to add RPATH would be good too)?

Reply:
Sure, I will do that. I have a question about this as well. The build system will pick up the flags from the postgresql pkg-config file, specifically, `.../postgresql-11.2-{HASH}/lib/pkgconfig/libpq.pc`. The value that it gets for my build is:

```
Libs: -L/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/postgresql-11.2-2cuj3k2jjbwkckphpkukjiwtd7ocmtbu/lib -lpq
```
Should the rpath flags be in the pkgconfig file for the postgresql package?

Reply:
If the R build system is using Spack's compiler wrappers there should be no issue as the RPATH is added transparently (since `postgresql` is a dependency). Otherwise we may need to add it. Can we check with either patchelf or obj dump if the RPATH is set correctly?

Reply:
The R build system uses Spack's compiler wrapper, and generally RPATH works fine. I have only seen this problem with r-rmariadb and r-rpostgres, both of which make use of pkg-config. It could be more prevalent though. Interestingly, the older r-rpostgresql, which does not use pkg-config, sets the RPATH for postgresql just fine.

Here is the information that you asked for:
```console
objdump -x opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgres-1.3.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw/rlib/R/library/RPostgres/libs/RPostgres.so | grep RPATH
  RPATH                /opt/ssoft/apps/2020.2/linux-centos7-x86_64/gcc-4.8.5/gcc-8.4.0-247uqcb/lib:/opt/ssoft/apps/2020.2/linux-centos7-x86_64/gcc-4.8.5/gcc-8.4.0-247uqcb/lib64:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-4.0.4-rqh7fgl5pvnxfh5fvze5cid6wa7swoak/rlib/R/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgres-1.3.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgres-1.3.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw/lib64:/opt/ssoft/apps/2020.2/linux-centos7-x86_64/gcc-4.8.5/gcc-8.4.0-247uqcb/lib/gcc/x86_64-pc-linux-gnu/8.4.0
```

For comparison, here is the result for the older r-rpostgresql package:
```console
objdump -x opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgresql-0.6-2-pwuzuhwut3hht3zkgzci4bgefpoh2prq/rlib/R/library/RPostgreSQL/libs/RPostgreSQL.so | grep RPATH
  RPATH                /opt/ssoft/apps/2020.2/linux-centos7-x86_64/gcc-4.8.5/gcc-8.4.0-247uqcb/lib:/opt/ssoft/apps/2020.2/linux-centos7-x86_64/gcc-4.8.5/gcc-8.4.0-247uqcb/lib64:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-4.0.3-yoggnajdqsblq6t5kqbtmwovzgye46st/rlib/R/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgresql-0.6-2-pwuzuhwut3hht3zkgzci4bgefpoh2prq/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgresql-0.6-2-pwuzuhwut3hht3zkgzci4bgefpoh2prq/lib64:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/ncurses-6.2-mb3uyhaol4nwz6hvabrt5agyz5ubut5w/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/openssl-1.1.1i-lrshpxps25g2ja75bu6cc7crga5uepdb/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/zlib-1.2.11-j4klyoyfqcyevegm644mqar4k7mbbq7m/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/postgresql-11.2-2cuj3k2jjbwkckphpkukjiwtd7ocmtbu/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/readline-8.0-w27vwxe7sv4r7o6n74g4n2z4fw354zhc/lib:/opt/ssoft/apps/2020.2/linux-centos7-x86_64/gcc-4.8.5/gcc-8.4.0-247uqcb/lib/gcc/x86_64-pc-linux-gnu/8.4.0
```

Reply:
The above was without the patch and the custom args. Here is the relevant output from the build log.
```
==> r-rpostgres: Executing phase: 'install'
==> [2021-03-23-20:35:46.626362] '/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-4.0.4-rqh7fgl5pvnxfh5fvze5cid6wa
7swoak/bin/R' '--vanilla' 'CMD' 'INSTALL' '--library=/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgres-1.3
.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw/rlib/R/library' '/tmp/gpjohnsn/spack-stage/spack-stage-r-rpostgres-1.3.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw
/spack-src'
* installing *source* package 'RPostgres' ...
** package 'RPostgres' successfully unpacked and MD5 sums checked
** using staged installation
Found pkg-config cflags and libs (libpq 11.2)!
...
Using PKG_CFLAGS=-I/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/postgresql-11.2-orjwzd3tyoiehupjzodiv24v4cjrtbyq/include
Using PKG_LIBS=-L/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/postgresql-11.2-orjwzd3tyoiehupjzodiv24v4cjrtbyq/lib -lpq
```

Reply:
@alalazo I looked into this more and am somewhat at a loss. As a refresher, r-rpostgresql is an older package and rpath works fine. The r-rpostgres package is newer and rpath does not work. I built r-rpostgresql and r-rpostgres with `--debug` and confirmed that rpath information is being set for r-rpostgresql but not r-rpostgres. Looking further, I can see that the reason for this is that `SPACK_*` variables are not getting set. For example, from the respective `spack-build-env.txt` files:

### r-rpostgresql
Transformed to make it easier to read.
```console
SPACK_RPATH_DIRS=
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgresql-0.6-2-lsjvvatmzdv5zw2mwl7pzikbgutdy7nb/lib:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgresql-0.6-2-lsjvvatmzdv5zw2mwl7pzikbgutdy7nb/lib64:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/postgresql-11.2-orjwzd3tyoiehupjzodiv24v4cjrtbyq/lib:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/zlib-1.2.11-j4klyoyfqcyevegm644mqar4k7mbbq7m/lib:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/readline-8.0-w27vwxe7sv4r7o6n74g4n2z4fw354zhc/lib:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/ncurses-6.2-mb3uyhaol4nwz6hvabrt5agyz5ubut5w/lib:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/openssl-1.1.1j-4223fvwzmochhtu32u3e7qrbrlrw4nqc/lib
```

### r-rpostgres
Transformed to make it easier to read.
```console
SPACK_RPATH_DIRS=
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgres-1.3.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw/lib:
/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/r-rpostgres-1.3.1-2kewd3pij6ima7je3r3k3xkpqjfzmmrw/lib64
```

Some differences to note between these two packages:

- r-rpostgresql uses automake configure; r-rpostgres uses a generic configure script
- r-rpostgresql configures postgresql from pg_config; r-rpostgres uses pkg-config
This does not seem like it should matter as they both set the same variables in Makevars.
- r-rpostgresql builds with cc; r-rpostgres builds with c++

Do you have any ideas as to why the `SPACK_*` variables are not getting the information from `set_build_environment_variables` for r-rpostgres?

Reply:
@glennpj I'll try to have a look this week

Reply:
I finally got back around to this and realized that I had made a copy/paste error, forgetting to remove the type attribute from the postgresql dependency, meaning it was not a link dependency.